File: github/app/src/main/java/com/einkreader/ui/reader/DebugLogActivity.java
package com.einkreader.ui.reader;

import android.app.Activity;
import android.os.Bundle;
import android.view.Gravity;
import android.view.View;
import android.widget.LinearLayout;
import android.widget.ScrollView;
import android.widget.TextView;

/**
 * 调试日志查看页面
 */
public class DebugLogActivity extends Activity {

    private TextView logView;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        LinearLayout root = new LinearLayout(this);
        root.setOrientation(LinearLayout.VERTICAL);
        root.setBackgroundColor(0xFFFFFFFF);

        // 标题栏
        LinearLayout titleBar = new LinearLayout(this);
        titleBar.setOrientation(LinearLayout.HORIZONTAL);
        titleBar.setBackgroundColor(0xFF333333);
        titleBar.setPadding(dp(16), dp(10), dp(16), dp(10));

        TextView btnBack = new TextView(this);
        btnBack.setText("← 返回");
        btnBack.setTextColor(0xFFFFFFFF);
        btnBack.setTextSize(18);
        btnBack.setPadding(dp(8), dp(6), dp(8), dp(6));
        btnBack.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) { finish(); }
        });
        titleBar.addView(btnBack);

        TextView tvTitle = new TextView(this);
        tvTitle.setText("调试日志");
        tvTitle.setTextColor(0xFFFFFFFF);
        tvTitle.setTextSize(20);
        tvTitle.setGravity(Gravity.CENTER);
        titleBar.addView(tvTitle, new LinearLayout.LayoutParams(0, -2, 1f));

        root.addView(titleBar, new LinearLayout.LayoutParams(-1, dp(50)));

        // 清空按钮
        TextView btnClear = new TextView(this);
        btnClear.setText("清空日志");
        btnClear.setTextColor(0xFF000000);
        btnClear.setTextSize(16);
        btnClear.setPadding(dp(16), dp(8), dp(16), dp(8));
        btnClear.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                DebugLog.clear();
                logView.setText("日志已清空\n");
            }
        });
        root.addView(btnClear);

        // 日志内容
        ScrollView scroll = new ScrollView(this);
        logView = new TextView(this);
        logView.setTextSize(12);
        logView.setTextColor(0xFF000000);
        logView.setTypeface(android.graphics.Typeface.MONOSPACE);
        logView.setPadding(dp(8), dp(8), dp(8), dp(8));
        scroll.addView(logView);

        root.addView(scroll, new LinearLayout.LayoutParams(-1, 0, 1f));

        setContentView(root);
    }

    @Override
    protected void onResume() {
        super.onResume();
        String log = DebugLog.getLog();
        StringBuilder sb = new StringBuilder();
        sb.append("【日志文件位置】\n").append(DebugLog.getLogFilePath()).append("\n");
        sb.append("(在文件管理器-下载 里可以找到)\n");
        sb.append("设备: " + android.os.Build.MODEL + " SDK:" + android.os.Build.VERSION.SDK_INT + " Android:" + android.os.Build.VERSION.RELEASE + "\n");
        sb.append("----\n\n");
        if (log.isEmpty()) {
            sb.append("(暂无日志)\n提示：打开一本书触发布局操作后日志才会出现");
        } else {
            sb.append(log);
        }
        logView.setText(sb.toString());
    }

    private int dp(int dp) {
        return (int)(dp * getResources().getDisplayMetrics().density + 0.5f);
    }
}

File: app/src/main/java/com/einkreader/ui/reader/DebugLogActivity.kt
package com.einkreader.ui.reader

import android.app.Activity
import android.graphics.Typeface
import android.os.Bundle
import android.view.Gravity
import android.view.View
import android.widget.LinearLayout
import android.widget.ScrollView
import android.widget.TextView

class DebugLogActivity : Activity() {
    private var logView: TextView? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        val root = LinearLayout(this).apply {
            orientation = LinearLayout.VERTICAL
            setBackgroundColor(-1) // 0xFFFFFFFF as signed
        }

        val titleBar = LinearLayout(this).apply {
            orientation = LinearLayout.HORIZONTAL
            setBackgroundColor(0x33333333.toInt())
            setPadding(dp(16), dp(10), dp(16), dp(10))

            val btnBack = TextView(this@DebugLogActivity).apply {
                text = "\u2190 \u8fd4\u56de"
                setTextColor(-1)
                textSize = 18f
                setPadding(dp(8), dp(6), dp(8), dp(6))
                setOnClickListener { finish() }
            }
            addView(btnBack)

            val tvTitle = TextView(this@DebugLogActivity).apply {
                text = "\u8c03\u8bd5\u65e5\u5fd7"
                setTextColor(-1)
                textSize = 20f
                gravity = Gravity.CENTER
            }
            addView(tvTitle, LinearLayout.LayoutParams(0, -2, 1f))
        }
        root.addView(titleBar, LinearLayout.LayoutParams(-1, dp(50)))

        val btnClear = TextView(this).apply {
            text = "\u6e05\u7a7a\u65e5\u5fd7"
            setTextColor((-0xffffff01).toInt())
            textSize = 16f
            setPadding(dp(16), dp(8), dp(16), dp(8))
            setOnClickListener {
                DebugLog.clear()
                logView?.setText("\u65e5\u5fd7\u5df2\u6e05\u7a7a\n")
            }
        }
        root.addView(btnClear)

        val scroll = ScrollView(this)
        logView = TextView(this).apply {
            textSize = 12f
            setTextColor((-0xffffff01).toInt())
            typeface = Typeface.MONOSPACE
            setPadding(dp(8), dp(8), dp(8), dp(8))
        }
        scroll.addView(logView!!)
        root.addView(scroll, LinearLayout.LayoutParams(-1, 0, 1f))

        setContentView(root)
    }

    override fun onResume() {
        super.onResume()
        val log = DebugLog.getLog()
        val sb = StringBuilder().apply {
            append("\u3010\u65e5\u5fd7\u6587\u4ef6\u4f4d\u7f6e\u3011\n")
            append(DebugLog.getLogFilePath()).append("\n")
            append("\uff08\u5728\u6587\u4ef6\u7ba1\u7406\u5668-\u4e0b\u8f7d \u91cc\u53ef\u4ee5\u627e\u5230\uff09\n")
            append("Device: ${android.os.Build.MODEL} SDK:${android.os.Build.VERSION.SDK_INT} Android:${android.os.Build.VERSION.RELEASE}\n")
            append("----\n\n")
            if (log.isEmpty()) {
                append("\uff08\u6682\u65e0\u65e5\u5fd7\uff09\n\u63d0\u793a\uff1a\u6253\u5f00\u4e00\u672c\u4e66\u89e6\u53d1\u5e03\u5c40\u64cd\u4f5c\u540e\u65e5\u5fd7\u624d\u4f1a\u51fa\u73b0")
            } else {
                append(log)
            }
        }
        logView?.text = sb.toString()
    }

    private fun dp(dpVal: Int): Int {
        return (dpVal * resources.displayMetrics.density + 0.5f).toInt()
    }
}
